### PR TITLE
login: T4943: use pam-auth-update to enable/disable Google authenticator

### PR DESCRIPTION
**Pull request by c-po**
## Change Summary

The initial version always enabled Google authenticator (2FA/MFA) support by hardcoding the PAM module for sshd and login.

This change only enables the PAM module on demand if any use has 2FA/MFA configured. Enabling the module is done system wide via pam-auth-update by using a predefined template.

Can be tested using:

```console
set system login user vyos authentication plaintext-password vyos
set system login user vyos authentication otp key 'QY735IG5HDHBFHS5W7Y2A4EM274SMT3O'
```

See https://docs.vyos.io/en/latest/configuration/system/login.html for additional details.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T4943

## Component(s) name

login

## Smoketest result

image

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: src/pam-configs/mfa-google-authenticator [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Name: Google Authenticator PAM module (2FA/MFA)
+Default: no
+Priority: 9999
```
Comment (zdc) [inline src/pam-configs/mfa-google-authenticator]: I think the priority should be `384` (based on https://wiki.ubuntu.com/PAMConfigFrameworkSpec)

`384` also will put it between `unix` and `radius/tacplus`, thus successful authentication using network PAM modules will skip MFA.

File: src/pam-configs/mfa-google-authenticator [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+Auth-Type: Primary
+Auth:
+    required      pam_google_authenticator.so nullok forward_pass
+
+Account-Type: Primary
+Account:
+
+Session-Type: Additional
+Session:
```
Comment (zdc) [inline src/pam-configs/mfa-google-authenticator]: Missing types can be omitted.

File: src/pam-configs/mfa-google-authenticator [hunk 3]
```diff
@@ -0,0 +1,13 @@
+Name: Google Authenticator PAM module (2FA/MFA)
+Default: no
+Priority: 9999
+
+Auth-Type: Primary
+Auth:
+    required      pam_google_authenticator.so nullok forward_pass
```
Comment (zdc) [inline src/pam-configs/mfa-google-authenticator]: This is equal to:

```
[default=bad success=ok new_authtok_reqd=ok ignore=ignore] pam_google_authenticator.so nullok forward_pass
```

But I would prefer:

```
[default=ignore success=ok auth_err=die] pam_google_authenticator.so nullok forward_pass
```

The module returns only `PAM_SUCCESS`, `PAM_IGNORE`, and `PAM_AUTH_ERR`. The main difference is that we will not call `unix` or other modules if the MFA code is wrong.

**Review by zdc — CHANGES_REQUESTED** [review]

File: src/pam-configs/mfa-google-authenticator (hunk 1 above)
Comment (c-po) [reply src/pam-configs/mfa-google-authenticator]: In the original implementation it was the first entry, that's why I re-added it at first position.
Happy to use a sane and reasonable position in the config

**Comment by c-po** [comment]
Incorporated requested changes by @zdc

**Comment by GurliGebis** [comment]
Would it make sense to rename it from "Google Authenticator" to just "Authenticator"?
Since you can use several different apps for it (Like Microsoft Authenticator, which a lot of people who has to use Office 365 already have installed) - it would make sense to remove the Google branding.

**Review by zdc — APPROVED** [review]

**Review by sever-sever — APPROVED** [review]

**Comment by c-po** [comment]
> Would it make sense to rename it from "Google Authenticator" to just "Authenticator"? Since you can use several different apps for it (Like Microsoft Authenticator, which a lot of people who has to use Office 365 already have installed) - it would make sense to remove the Google branding.

Well it was invented by Google that should be honored. I use it with Microsoft authenticator, too.

**Comment by c-po** [comment]
@Mergifyio backport sagitta

**Comment by GurliGebis** [comment]
@c-po that makes sense.
Just as long as people aren't misled into thinking they only need to use the Google app 🙂

**Pull request merged**